<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro"  name="tennis_ball">

  <!-- <xacro:include filename="$(find xarm_description)/urdf/xarm_device.urdf.xacro" /> -->

  <!-- Material Definition -->
  <material name="yellow">
    <color rgba="1 1 0 1"/>
  </material>

  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>

  <link name="world">
  </link>

  <link name="base_link">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <visual>
      <geometry>
        <box size="1.2 0.01 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <box size="1.2 0.01 0.01"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.008333333333" ixy="0" ixz="0" iyy="0.0.06041666667" iyz="0" izz="0.0.06041666667"/>
    </inertial>
  </link>

  <!-- The cylindrical robot -->
  <link name="theta_link">
    <visual>
      <origin xyz="0 0.380 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.75 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.75 0.01"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0.06041666667" ixy="0" ixz="0" iyy="0.0008333333333" iyz="0" izz="0.0.06041666667"/>
    </inertial>
  </link>
  
  
  <link name="r_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 2.0"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 2.0"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0.06041666667" ixy="0" ixz="0" iyy="0.0.06041666667" iyz="0" izz="0.0008333333333"/>
    </inertial>
  </link>

  
  <link name="z_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0.008333333333" ixy="0" ixz="0" iyy="0.0008333333333" iyz="0" izz="0.0.008333333333"/>
    </inertial>
  </link>
  
  <!-- Link Definition -->
  <link name="ball">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.033"/>
      </geometry>
      <material name="yellow"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.033"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  
  <!-- Joints  -->
  <joint name="world_to_base_link_joint" type="fixed">
      <origin xyz="-0.2 -0.415 0.005" rpy="0 0 0"/>
      <parent link="world"/>
      <child link="base_link"/>
  </joint>

  <joint name="theta_joint" type="revolute">
      <axis xyz="0 0 1"/>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="theta_link"/>
      <limit effort="50" velocity="0.01" lower="-1.5707" upper="1.5707"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="r_joint" type="prismatic">
      <origin xyz="0 0.505 1.005" rpy="0 0 0"/>
      <parent link="theta_link"/>
      <child link="r_link"/>
      <axis xyz="0 1 0"/>
      <limit effort="50" velocity="0.01" lower="-0.35" upper="0.1"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="z_joint" type="prismatic">
      <origin xyz="0 -0.005 0" rpy="0 0 0"/>
      <parent link="r_link"/>
      <child link="z_link"/>
      <axis xyz="0 0 1"/>
      <limit effort="50" velocity="0.01" lower="-0.2" upper="0.75"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="fixed_joint" type="fixed">
      <origin xyz="0 -0.033 0" rpy="0 0 0"/>
      <parent link="z_link"/>
      <child link="ball"/>
  </joint>

  <xacro:include filename="gazebo_plugin_cylindrical.xacro" />


</robot>


<!-- ros2 topic pub -1 /set_joint_trajectory trajectory_msgs/msg/JointTrajectory  '{header: {frame_id: world}, joint_names: [theta_joint, r_joint, z_joint], points: [{positions: {1.0,0.3,0.3}]}' -->